integration-webcert/src/main/java/se/inera/intyg/minaintyg/integration/webcert/converter/data/CertificateElementOrganizer.java

```diff
@@ -0,0 +1,82 @@
+package se.inera.intyg.minaintyg.integration.webcert.converter.data;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Component;
+import se.inera.intyg.minaintyg.integration.webcert.client.dto.CertificateDataElement;
+import se.inera.intyg.minaintyg.integration.webcert.client.dto.config.CertificateDataConfigTypes;
+
+@Component
+public class CertificateElementOrganizer {
+
+  public Map<CertificateDataElement, List<CertificateDataElement>> organize(
+      List<CertificateDataElement> certificateDataElements) {
+
+    final var categoryMap = getCategoryMap(certificateDataElements);
+    final var questionMap = getQuestionMap(certificateDataElements);
+
+    final var categoryQuestionMap = categoryMap.values().stream()
+        .flatMap(category ->
+            questionMap.values().stream()
+                .filter(questionBelongToCategory(category, questionMap))
+                .map(question -> new SimpleEntry<>(category, question))
+        )
+        .collect(
+            Collectors.groupingBy(
+                SimpleEntry::getKey,
+                Collectors.mapping(
+                    SimpleEntry::getValue,
+                    Collectors.toList()
+                )
+            )
+        );
+
+    sortQuestionsInIndexOrder(categoryQuestionMap);
+
+    return categoryQuestionMap;
+  }
+
+  private static Map<String, CertificateDataElement> getQuestionMap(
+      List<CertificateDataElement> certificateDataElements) {
+    return certificateDataElements.stream()
+        .filter(Predicate.not(CertificateElementOrganizer::elementIsCategory))
```

Comment:
getQuestionMap and getCategoryMap are almost the same. Could they use common logic, or do you feel like that would make the code harder to read?

Reply:
I decide for now to leave as is... the reason is for simplicity and readability.